File: source/milsko/milsko.cpp
#include <os.hpp>
#include <runtime/blocks/blockUtils.hpp>

#define MW_OPENGL_NO_INCLUDE
#define MW_VULKAN_NO_INCLUDE
#include <Mw/Milsko.h>
#include <Mw/Widget/OpenGL.h>
#include <Mw/Widget/Vulkan.h>
#include <stdio.h>

static void* to_ptr(Value src){
	void* dest;

	sscanf(src.asString().c_str(), "%p", (void**)&dest);

	return dest;
}

static std::string from_ptr(void* ptr){
	char cptr[64];

	sprintf(cptr, "%p", ptr);

	return std::string(cptr);
}

SCRATCH_BLOCK(milsko, init) {
	Log::log("Milsko initialization");
	MwLibraryInit();
	return BlockResult::CONTINUE;
}

SCRATCH_BLOCK(milsko, step) {
	Value widgetValue;
	MwWidget widget;
	if(!Scratch::getInput(block, "WIDGET", thread, sprite, widgetValue)) return BlockResult::REPEAT;

	widget = (MwWidget)to_ptr(widgetValue);

	*outValue = Value(MwStep(widget) == 0);
	return BlockResult::CONTINUE;
}

SCRATCH_BLOCK(milsko, pending) {
	Value widgetValue;
	MwWidget widget;
	if(!Scratch::getInput(block, "WIDGET", thread, sprite, widgetValue)) return BlockResult::REPEAT;

	widget = (MwWidget)to_ptr(widgetValue);

	*outValue = Value(!!MwPending(widget));
	return BlockResult::CONTINUE;
}

SCRATCH_BLOCK(milsko, none) {
	*outValue = Value(from_ptr(NULL));
	return BlockResult::CONTINUE;
}

SCRATCH_BLOCK(milsko, default) {
	*outValue = Value(MwDEFAULT);
	return BlockResult::CONTINUE;
}

SCRATCH_BLOCK(milsko, destroy) {
	Value widgetValue;
	MwWidget widget;
	if(!Scratch::getInput(block, "WIDGET", thread, sprite, widgetValue)) return BlockResult::REPEAT;

	widget = (MwWidget)to_ptr(widgetValue);
	MwDestroyWidget(widget);

	return BlockResult::CONTINUE;
}

SCRATCH_BLOCK(milsko, move) {
	Value widgetValue, xValue, yValue;
	MwWidget widget;
	if(!Scratch::getInput(block, "WIDGET", thread, sprite, widgetValue)) return BlockResult::REPEAT;
	if(!Scratch::getInput(block, "X", thread, sprite, xValue)) return BlockResult::REPEAT;
	if(!Scratch::getInput(block, "Y", thread, sprite, yValue)) return BlockResult::REPEAT;

	widget = (MwWidget)to_ptr(widgetValue);
	MwVaApply(widget,
		MwNx, (int)xValue.asDouble(),
		MwNy, (int)yValue.asDouble(),
	NULL);

	return BlockResult::CONTINUE;
}

SCRATCH_BLOCK(milsko, resize) {
	Value widgetValue, widthValue, heightValue;
	MwWidget widget;
	if(!Scratch::getInput(block, "WIDGET", thread, sprite, widgetValue)) return BlockResult::REPEAT;
	if(!Scratch::getInput(block, "WIDTH", thread, sprite, widthValue)) return BlockResult::REPEAT;
	if(!Scratch::getInput(block, "HEIGHT", thread, sprite, heightValue)) return BlockResult::REPEAT;

	widget = (MwWidget)to_ptr(widgetValue);
	MwVaApply(widget,
		MwNwidth, (int)widthValue.asDouble(),
		MwNheight, (int)heightValue.asDouble(),
	NULL);

	return BlockResult::CONTINUE;
}

SCRATCH_SHADOW_BLOCK(milsko_menu_WIDGET_CLASS, WIDGET_CLASS);

#define TYPE(type, mtype, stype, prefix, stuff, stuff2) \
SCRATCH_SHADOW_BLOCK(milsko_menu_ ## stype ## _PROP, stype ## _PROP); \
SCRATCH_BLOCK(milsko, set ## type) { \
	Value widgetValue, propValue, valueValue; \
	MwWidget widget; \
	if(!Scratch::getInput(block, "WIDGET", thread, sprite, widgetValue)) return BlockResult::REPEAT; \
	if(!Scratch::getInput(block, "PROP", thread, sprite, propValue)) return BlockResult::REPEAT; \
	if(!Scratch::getInput(block, "VALUE", thread, sprite, valueValue)) return BlockResult::REPEAT; \
\
	widget = (MwWidget)to_ptr(widgetValue); \
	MwSet ## mtype(widget, (prefix + propValue.asString()).c_str(), stuff); \
\
	return BlockResult::CONTINUE; \
} \
SCRATCH_BLOCK(milsko, get ## type) { \
	Value widgetValue, propValue; \
	MwWidget widget; \
	if(!Scratch::getInput(block, "WIDGET", thread, sprite, widgetValue)) return BlockResult::REPEAT; \
	if(!Scratch::getInput(block, "PROP", thread, sprite, propValue)) return BlockResult::REPEAT; \
\
	widget = (MwWidget)to_ptr(widgetValue); \
\
	*outValue = Value(stuff2(MwGet ## mtype(widget, (prefix + propValue.asString()).c_str()))); \
	return BlockResult::CONTINUE; \
}

TYPE(Integer, Integer, INTEGER, "I", (int)valueValue.asDouble(), (int));
TYPE(String, Text, STRING, "S", valueValue.asString().c_str(), std::string);
TYPE(Void, Void, VOID, "V", to_ptr(valueValue), from_ptr);

SCRATCH_BLOCK(milsko, create) {
	Value widgetClassValue, nameValue, xValue, yValue, widthValue, heightValue, parentValue;
	MwWidget widget, parent;
	MwClass widget_class = NULL;
	if(!Scratch::getInput(block, "WIDGET_CLASS", thread, sprite, widgetClassValue)) return BlockResult::REPEAT;
	if(!Scratch::getInput(block, "NAME", thread, sprite, nameValue)) return BlockResult::REPEAT;
	if(!Scratch::getInput(block, "X", thread, sprite, xValue)) return BlockResult::REPEAT;
	if(!Scratch::getInput(block, "Y", thread, sprite, yValue)) return BlockResult::REPEAT;
	if(!Scratch::getInput(block, "WIDTH", thread, sprite, widthValue)) return BlockResult::REPEAT;
	if(!Scratch::getInput(block, "HEIGHT", thread, sprite, heightValue)) return BlockResult::REPEAT;
	if(!Scratch::getInput(block, "PARENT", thread, sprite, parentValue)) return BlockResult::REPEAT;

#define WIDGET(x) \
	if(widgetClassValue.asString() == #x){ \
		widget_class = Mw ## x ## Class; \
	}
	
/* BEGIN WIDGETS */
	WIDGET(Box);
	WIDGET(Button);
	WIDGET(CheckBox);
	WIDGET(Entry);
	WIDGET(Frame);
	WIDGET(Image);
	WIDGET(Label);
	WIDGET(ListBox);
	WIDGET(Menu);
	WIDGET(NumberEntry);
	WIDGET(ScrollBar);
	WIDGET(SubMenu);
	WIDGET(Viewport);
	WIDGET(Window);
	WIDGET(ProgressBar);
	WIDGET(RadioBox);
	WIDGET(ComboBox);
	WIDGET(TreeView);
	WIDGET(OpenGL);
	WIDGET(Vulkan);
/* END WIDGETS */

#undef WIDGET

	parent = to_ptr(parentValue);
	
	widget = MwCreateWidget(widget_class, nameValue.asString().c_str(), parent, (int)(xValue.asDouble()), (int)(yValue.asDouble()), (unsigned int)(widthValue.asDouble()), (unsigned int)(heightValue.asDouble()));
	
	*outValue = Value(from_ptr((void*)widget));
	return BlockResult::CONTINUE;
}

File: source/milsko/milsko.js
(function (Scratch) {
 	let widgets = [
		/* BEGIN WIDGETS */
		"Box", "Button", "CheckBox", "Entry", "Frame", "Image", "Label", "ListBox", "Menu", "NumberEntry", "ScrollBar", "SubMenu", "Viewport", "Window", "ProgressBar", "RadioBox", "ComboBox", "TreeView", "OpenGL", "Vulkan"
		/* END WIDGETS */
	];

	class Milsko {
		getInfo() {
			return {
				id: "milsko",
				name: "Milsko",
				menus: {
					INTEGER_PROP: {
						acceptReporters: true,
						items: [
							/* BEGIN INTEGER PROPS */
							"x", "y", "width", "height", "orientation", "minValue", "maxValue", "value", "changedBy", "areaShown", "checked", "alignment", "bold", "main", "leftPadding", "hasHeading", "hasBorder", "inverted", "modernLook", "hideInput", "singleClickSelectable", "flat", "showArrows", "padding", "borderWidth", "ratio", "fixedSize", "margin", "bitmapFont", "sevenSegment", "length", "forceInverted", "isRounded", "darkTheme", "useMonospace", "fillArea", "waitMS"
							/* END INTEGER PROPS */
						]
					},
					STRING_PROP: {
						acceptReporters: true,
						items: [
							/* BEGIN STRING PROPS */
							"title", "text", "background", "subBackground", "foreground", "subForeground", "vulkanExtension", "vulkanLayer"
							/* END STRING PROPS */
						]
					},
					VOID_PROP: {
						acceptReporters: true,
						items: [
							/* BEGIN VOID PROPS */
							"pixmap", "iconPixmap", "sizeHints", "font", "boldFont", "backgroundPixmap", "monospaceFont", "boldMonospaceFont", "vulkanConfig"
							/* END VOID PROPS */
						]
					},
					WIDGET_CLASS: {
						acceptReporters: false,
						items: widgets
					}
				},
				blocks: [
					{
						opcode: "init",
						blockType: Scratch.BlockType.COMMAND,
						text: "Initialize Milsko"
					},
					{
						opcode: "step",
						blockType: Scratch.BlockType.BOOLEAN,
						text: "Ran a single step for [WIDGET] and widgets want to keep running?",
						arguments: {
							WIDGET: {
								type: Scratch.ArgumentType.STRING
							}
						}
					},
					{
						opcode: "pending",
						blockType: Scratch.BlockType.BOOLEAN,
						text: "Any pending events for [WIDGET]?",
						arguments: {
							WIDGET: {
								type: Scratch.ArgumentType.STRING
							}
						}
					},
					{
						opcode: "none",
						blockType: Scratch.BlockType.REPORTER,
						text: "None"
					},
					{
						opcode: "default",
						blockType: Scratch.BlockType.REPORTER,
						text: "Default"
					},
					{
						opcode: "move",
						blockType: Scratch.BlockType.COMMAND,
						text: "Move widget [WIDGET] to ([X], [Y])",
						arguments: {
							WIDGET: {
								type: Scratch.ArgumentType.STRING
							},
							X: {
								type: Scratch.ArgumentType.NUMBER
							},
							Y: {
								type: Scratch.ArgumentType.NUMBER
							}
						}
					},
					{
						opcode: "resize",
						blockType: Scratch.BlockType.COMMAND,
						text: "Resize widget [WIDGET] to [WIDTH]x[HEIGHT]",
						arguments: {
							WIDGET: {
								type: Scratch.ArgumentType.STRING
							},
							WIDTH: {
								type: Scratch.ArgumentType.NUMBER
							},
							HEIGHT: {
								type: Scratch.ArgumentType.NUMBER
							}
						}
					},
					{
						opcode: "setInteger",
						blockType: Scratch.BlockType.COMMAND,
						text: "Set integer property [PROP] of widget [WIDGET] to [VALUE]",
						arguments: {
							WIDGET: {
								type: Scratch.ArgumentType.STRING
							},
							PROP: {
								type: Scratch.ArgumentType.STRING,
								menu: "INTEGER_PROP"
							},
							VALUE: {
								type: Scratch.ArgumentType.NUMBER
							}
						}
					},
					{
						opcode: "setString",
						blockType: Scratch.BlockType.COMMAND,
						text: "Set string property [PROP] of widget [WIDGET] to [VALUE]",
						arguments: {
							WIDGET: {
								type: Scratch.ArgumentType.STRING
							},
							PROP: {
								type: Scratch.ArgumentType.STRING,
								menu: "STRING_PROP"
							},
							VALUE: {
								type: Scratch.ArgumentType.STRING
							}
						}
					},
					{
						opcode: "setVoid",
						blockType: Scratch.BlockType.COMMAND,
						text: "Set void property [PROP] of widget [WIDGET] to [VALUE]",
						arguments: {
							WIDGET: {
								type: Scratch.ArgumentType.STRING
							},
							PROP: {
								type: Scratch.ArgumentType.STRING,
								menu: "VOID_PROP"
							},
							VALUE: {
								type: Scratch.ArgumentType.STRING
							}
						}
					},
					{
						opcode: "create",
						blockType: Scratch.BlockType.REPORTER,
						text: "Creates a [WIDGET_CLASS] widget named [NAME] at ([X], [Y]) with size [WIDTH]x[HEIGHT] as the child of [PARENT]",
						arguments: {
							WIDGET_CLASS: {
								type: Scratch.ArgumentType.STRING,
								menu: "WIDGET_CLASS",
							},
							NAME: {
								type: Scratch.ArgumentType.STRING
							},
							X: {
								type: Scratch.ArgumentType.NUMBER
							},
							Y: {
								type: Scratch.ArgumentType.NUMBER
							},
							WIDTH: {
								type: Scratch.ArgumentType.NUMBER
							},
							HEIGHT: {
								type: Scratch.ArgumentType.NUMBER
							},
							PARENT: {
								type: Scratch.ArgumentType.STRING
							}
						}
					}
				]
			};
		}

		init() {
		}

		step() {
		}

		pending() {
		}

		setInteger() {
		}

		setText() {
		}

		setVoid() {
		}

		create() {
		}
	};

	Scratch.extensions.register(new Milsko());
})(Scratch);
